From 6dd1f42038f5b599345a790ea51cb2ddbc6da982 Mon Sep 17 00:00:00 2001
From: Ivan Evlogiev <quic_ievlogie@quicinc.com>
Date: Mon, 20 Dec 2021 09:36:42 +0200
Subject: [PATCH] nn-framework: Add cmake option to enable/disable CPU
 implementation

- Add cmake option NN_INCLUDE_CPU_IMPLEMENTATION to enable
cpu implementation in nn framework
- By default cpu implementation is disabled

Change-Id: I64e737da38850da1a17f03eb933f86a451656887
---
 nn/CMakeLists.txt                                |  9 ++++++++-
 nn/cast/test_tools/vts-cros/CMakeLists.txt       | 15 ++++++++++++---
 nn/common/CMakeLists.txt                         |  9 +++++++--
 nn/common/CpuExecutor.cpp                        | 13 +++++++++++++
 nn/common/OperationsUtils.cpp                    |  2 ++
 nn/common/include/CpuOperationUtils.h            |  4 ++++
 nn/common/operations/Activation.cpp              |  6 ++++++
 nn/common/operations/ArgMinMax.cpp               |  2 ++
 .../operations/BidirectionalSequenceLSTM.cpp     |  2 ++
 nn/common/operations/BidirectionalSequenceLSTM.h |  4 ++++
 .../operations/BidirectionalSequenceRNN.cpp      |  4 ++++
 nn/common/operations/Broadcast.cpp               |  6 ++++++
 nn/common/operations/Cast.cpp                    |  2 ++
 nn/common/operations/Cast.h                      |  2 ++
 nn/common/operations/ChannelShuffle.cpp          |  4 ++++
 nn/common/operations/Comparisons.cpp             |  4 ++++
 nn/common/operations/Concatenation.cpp           |  6 ++++++
 nn/common/operations/Conv2D.cpp                  |  6 ++++++
 nn/common/operations/DepthwiseConv2D.cpp         |  2 ++
 nn/common/operations/Dequantize.cpp              |  4 ++++
 nn/common/operations/Elementwise.cpp             |  4 ++++
 nn/common/operations/EmbeddingLookup.cpp         |  2 ++
 nn/common/operations/EmbeddingLookup.h           |  2 ++
 nn/common/operations/EmbeddingLookupTest.cpp     |  2 ++
 nn/common/operations/ExpandDims.cpp              |  2 ++
 nn/common/operations/ExpandDims.h                |  2 ++
 nn/common/operations/FullyConnected.cpp          |  6 ++++++
 nn/common/operations/Gather.cpp                  |  4 ++++
 nn/common/operations/GenerateProposals.cpp       | 16 ++++++++++++++++
 nn/common/operations/GroupedConv2D.cpp           |  2 ++
 nn/common/operations/HalOperation.h              |  2 ++
 nn/common/operations/HashtableLookup.cpp         |  2 ++
 nn/common/operations/HashtableLookup.h           |  2 ++
 nn/common/operations/HashtableLookupTest.cpp     |  2 ++
 nn/common/operations/HeatmapMaxKeypoint.cpp      |  4 ++++
 nn/common/operations/InstanceNormalization.cpp   |  4 ++++
 nn/common/operations/L2Normalization.cpp         |  6 ++++++
 nn/common/operations/LSHProjection.cpp           |  2 ++
 nn/common/operations/LSHProjection.h             |  2 ++
 nn/common/operations/LSHProjectionTest.cpp       |  2 ++
 nn/common/operations/LSTM.cpp                    |  2 ++
 nn/common/operations/LSTM.h                      |  2 ++
 nn/common/operations/LSTMTest.cpp                |  2 ++
 nn/common/operations/LayerNormLSTMTest.cpp       |  3 +++
 nn/common/operations/LogSoftmax.cpp              |  4 ++++
 nn/common/operations/LogicalAndOr.cpp            |  4 ++++
 nn/common/operations/LogicalNot.cpp              |  4 ++++
 nn/common/operations/MaximumMinimum.cpp          |  2 ++
 nn/common/operations/MaximumMinimum.h            |  2 ++
 nn/common/operations/Multinomial.cpp             |  2 ++
 nn/common/operations/Multinomial.h               |  2 ++
 nn/common/operations/MultinomialTest.cpp         |  2 ++
 nn/common/operations/Neg.cpp                     |  4 ++++
 nn/common/operations/Normalization.cpp           |  2 ++
 nn/common/operations/PRelu.cpp                   |  6 ++++++
 nn/common/operations/Pooling.cpp                 |  6 ++++++
 nn/common/operations/Pow.cpp                     |  2 ++
 nn/common/operations/Pow.h                       |  2 ++
 nn/common/operations/Quantize.cpp                |  4 ++++
 nn/common/operations/QuantizedLSTM.cpp           |  2 ++
 nn/common/operations/QuantizedLSTM.h             |  2 ++
 nn/common/operations/QuantizedLSTMTest.cpp       |  2 ++
 nn/common/operations/RNN.cpp                     |  2 ++
 nn/common/operations/RNN.h                       |  2 ++
 nn/common/operations/RNNTest.cpp                 |  2 ++
 nn/common/operations/Reduce.cpp                  |  4 ++++
 nn/common/operations/Reshape.cpp                 |  2 ++
 nn/common/operations/ResizeImageOps.cpp          |  6 ++++++
 nn/common/operations/RoiAlign.cpp                |  6 ++++++
 nn/common/operations/RoiPooling.cpp              |  4 ++++
 nn/common/operations/SVDF.cpp                    |  2 ++
 nn/common/operations/SVDF.h                      |  3 ++-
 nn/common/operations/SVDFTest.cpp                |  2 ++
 nn/common/operations/Select.cpp                  |  4 ++++
 nn/common/operations/SimpleMath.cpp              |  2 ++
 nn/common/operations/Slice.cpp                   |  4 ++++
 nn/common/operations/Softmax.cpp                 |  6 ++++++
 nn/common/operations/Split.cpp                   |  2 ++
 nn/common/operations/StridedSlice.cpp            |  2 ++
 nn/common/operations/Tile.cpp                    |  2 ++
 nn/common/operations/Tile.h                      |  2 ++
 nn/common/operations/TopK_V2.cpp                 |  3 +++
 nn/common/operations/TopK_V2.h                   |  2 ++
 nn/common/operations/Transpose.cpp               |  6 ++++++
 nn/common/operations/TransposeConv2D.cpp         |  6 ++++++
 .../operations/UnidirectionalSequenceLSTM.cpp    |  4 ++++
 .../operations/UnidirectionalSequenceRNN.cpp     |  4 ++++
 87 files changed, 316 insertions(+), 7 deletions(-)

diff --git a/nn/CMakeLists.txt b/nn/CMakeLists.txt
index 38ad7ce3..d9c76fd8 100644
--- a/nn/CMakeLists.txt
+++ b/nn/CMakeLists.txt
@@ -15,10 +15,17 @@ add_definitions(-flax-vector-conversions)
 add_definitions(-fomit-frame-pointer)
 
 include_directories(${SYSROOT_INCDIR})
-include_directories(${SYSROOT_INCDIR}/tflite)
 include_directories(${SYSROOT_INCDIR}/eigen3)
 link_directories(${SYSROOT_LIBDIR})
 
+option(NN_INCLUDE_CPU_IMPLEMENTATION "Include CPU Implementation" OFF)
+
+if (NN_INCLUDE_CPU_IMPLEMENTATION)
+  include_directories(${SYSROOT_INCDIR}/tensorflow)
+  find_package(PkgConfig)
+  pkg_check_modules(TFLITE REQUIRED tensorflow-lite)
+endif ()
+
 add_subdirectory(cast/include)
 add_subdirectory(cast/android/hardware/neuralnetworks/1.2)
 add_subdirectory(cast/hidl_base_shlib)
diff --git a/nn/cast/test_tools/vts-cros/CMakeLists.txt b/nn/cast/test_tools/vts-cros/CMakeLists.txt
index f54a1cc7..da7114f2 100644
--- a/nn/cast/test_tools/vts-cros/CMakeLists.txt
+++ b/nn/cast/test_tools/vts-cros/CMakeLists.txt
@@ -58,11 +58,14 @@ target_link_libraries(${NNHAL_TEST_1_0_TARGET_NAME}
     vts_hal_hidl_target_test_base
     gtest
     gtest_main
-    tensorflow-lite
     base
     log
 )
 
+if (NN_INCLUDE_CPU_IMPLEMENTATION)
+    target_link_libraries(${NNHAL_TEST_1_0_TARGET_NAME} ${TFLITE_LIBRARIES})
+endif ()
+
 install(
     TARGETS ${NNHAL_TEST_1_0_TARGET_NAME}
     RUNTIME DESTINATION bin
@@ -97,11 +100,14 @@ target_link_libraries(${NNHAL_TEST_1_1_TARGET_NAME}
     vts_hal_hidl_target_test_base
     gtest
     gtest_main
-    tensorflow-lite
     base
     log
 )
 
+if (NN_INCLUDE_CPU_IMPLEMENTATION)
+    target_link_libraries(${NNHAL_TEST_1_1_TARGET_NAME} ${TFLITE_LIBRARIES})
+endif ()
+
 install(
     TARGETS ${NNHAL_TEST_1_1_TARGET_NAME}
     RUNTIME DESTINATION bin
@@ -144,11 +150,14 @@ target_link_libraries(${NNHAL_TEST_1_2_TARGET_NAME}
     vts_hal_hidl_target_test_base
     gtest
     gtest_main
-    tensorflow-lite
     base
     log
 )
 
+if (NN_INCLUDE_CPU_IMPLEMENTATION)
+    target_link_libraries(${NNHAL_TEST_1_2_TARGET_NAME} ${TFLITE_LIBRARIES})
+endif ()
+
 install(
     TARGETS ${NNHAL_TEST_1_2_TARGET_NAME}
     RUNTIME DESTINATION bin
diff --git a/nn/common/CMakeLists.txt b/nn/common/CMakeLists.txt
index 473736ad..c6223e3e 100644
--- a/nn/common/CMakeLists.txt
+++ b/nn/common/CMakeLists.txt
@@ -1,6 +1,8 @@
 set (TARGET_NAME neuralnetworks_common)
 
-add_definitions(-DNN_INCLUDE_CPU_IMPLEMENTATION)
+if (NN_INCLUDE_CPU_IMPLEMENTATION)
+    add_definitions(-DNN_INCLUDE_CPU_IMPLEMENTATION)
+endif ()
 
 set (SOURCE_FILES
     OperationResolver.cpp
@@ -96,13 +98,16 @@ target_link_libraries(${TARGET_NAME}
     textclassifier_hash
     cast_nn_hidl_memory
     cast_nn_hidl_allocator
-    tensorflow-lite
     philox_random
     android.hardware.neuralnetworks_1_2
     fmq
     crypto
 )
 
+if (NN_INCLUDE_CPU_IMPLEMENTATION)
+    target_link_libraries(${TARGET_NAME} ${TFLITE_LIBRARIES})
+endif ()
+
 install(
     TARGETS ${TARGET_NAME}
     LIBRARY DESTINATION ${NNAPI_LIB_DIR}
diff --git a/nn/common/CpuExecutor.cpp b/nn/common/CpuExecutor.cpp
index 86835b80..7d9b227c 100644
--- a/nn/common/CpuExecutor.cpp
+++ b/nn/common/CpuExecutor.cpp
@@ -598,6 +598,11 @@ int CpuExecutor::run(const Model& model, const Request& request,
     NNTRACE_CPU(NNTRACE_PHASE_EXECUTION, "run");
     VLOG(CPUEXE) << "CpuExecutor::run() with request(" << SHOW_IF_DEBUG(toString(request)) << ")";
 
+#ifndef NN_INCLUDE_CPU_IMPLEMENTATION
+    LOG(ERROR) << "CPU implementation is disabled";
+    return ANEURALNETWORKS_UNAVAILABLE_DEVICE;
+#endif
+
     // b/109953668, disable OpenMP
 #ifdef NNAPI_OPENMP
     ScopedOpenmpSettings openMpSettings;
@@ -727,6 +732,7 @@ void CpuExecutor::freeNoLongerUsedOperands(const std::vector<uint32_t>& inputs)
     }
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 int CpuExecutor::executeOperation(const Operation& operation) {
     // VLOG(CPUEXE) << "CpuExecutor::executeOperation(" << toString(operation) << ")";
     const hidl_vec<uint32_t>& ins = operation.inputs;
@@ -1822,6 +1828,13 @@ int CpuExecutor::executeOperation(const Operation& operation) {
     freeNoLongerUsedOperands(ins);
     return ANEURALNETWORKS_NO_ERROR;
 }
+#else
+int CpuExecutor::executeOperation(const Operation& operation) {
+    (void)(operation);
+    LOG(ERROR) << "CPU implementation is disabled";
+    return ANEURALNETWORKS_UNAVAILABLE_DEVICE;
+}
+#endif
 
 void CpuExecutor::finish(int result) {
     // Free allocated temporary operands.
diff --git a/nn/common/OperationsUtils.cpp b/nn/common/OperationsUtils.cpp
index e128afbb..7bea7f9f 100644
--- a/nn/common/OperationsUtils.cpp
+++ b/nn/common/OperationsUtils.cpp
@@ -225,6 +225,7 @@ bool GetQuantizedConvolutionMultipler(const Shape& inputShape, const Shape& filt
     return true;
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 void CalculateActivationRangeUint8(int32_t activation,
                                    const Shape& outputShape,
                                    int32_t* act_min,
@@ -275,6 +276,7 @@ void CalculateActivationRangeFloat(int32_t activation,
         LOG(ERROR) << "Unsupported fused activation function.";
     }
 }
+#endif
 
 int32_t CalculateInputRadius(int input_integer_bits, int input_left_shift) {
     const double max_input_rescaled = 1.0 * ((1 << input_integer_bits) - 1) *
diff --git a/nn/common/include/CpuOperationUtils.h b/nn/common/include/CpuOperationUtils.h
index c0dd00c3..e6011ee2 100644
--- a/nn/common/include/CpuOperationUtils.h
+++ b/nn/common/include/CpuOperationUtils.h
@@ -23,13 +23,16 @@
 #include <cmath>
 #include <limits>
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/types.h"
+#endif
 
 namespace android {
 namespace nn {
 
 // The implementations in tflite/kernels/internal/ take a Dims<4> object
 // even if the original tensors were not 4D.
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 inline tflite::Dims<4> convertShapeToDims(const Shape& shape) {
     nnAssert(shape.dimensions.size() <= 4);
     tflite::Dims<4> dims;
@@ -57,6 +60,7 @@ inline tflite::RuntimeShape convertShapeToTflshape(const Shape& shape) {
     std::vector<int32_t> tflShapeDim(shape.dimensions.begin(), shape.dimensions.end());
     return tflite::RuntimeShape(tflShapeDim.size(), tflShapeDim.data());
 }
+#endif
 
 inline void convertFloat16ToFloat32(const _Float16* input, std::vector<float>* output) {
     CHECK(input != nullptr);
diff --git a/nn/common/operations/Activation.cpp b/nn/common/operations/Activation.cpp
index f85f6b4b..8cd5c645 100644
--- a/nn/common/operations/Activation.cpp
+++ b/nn/common/operations/Activation.cpp
@@ -18,8 +18,10 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
 #include "tensorflow/lite/kernels/internal/optimized/optimized_ops.h"
+#endif
 
 #include "Tracing.h"
 
@@ -34,6 +36,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -205,6 +208,7 @@ bool logisticQuant8(const uint8_t* inputData, const Shape& inputShape, uint8_t*
 }
 
 }  // namespace
+#endif
 
 bool validate(OperationType opType, const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -226,6 +230,7 @@ bool validate(OperationType opType, const IOperationValidationContext* context)
     return validateInputTypes(context, {inputType}) && validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(OperationType opType, IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     NN_RET_CHECK_LE(getNumberOfDimensions(input), 4);
@@ -370,6 +375,7 @@ bool executeTanh(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation TANH";
     }
 }
+#endif
 
 }  // namespace activation
 
diff --git a/nn/common/operations/ArgMinMax.cpp b/nn/common/operations/ArgMinMax.cpp
index b323d517..de4d599a 100644
--- a/nn/common/operations/ArgMinMax.cpp
+++ b/nn/common/operations/ArgMinMax.cpp
@@ -16,6 +16,7 @@
 
 // Contains the implementation of the operations.
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "Operations.h"
@@ -83,3 +84,4 @@ bool argMinMaxGeneric(const uint8_t* inputData, const Shape& inputShape,
 
 } // namespace nn
 } // namespace android
+#endif
\ No newline at end of file
diff --git a/nn/common/operations/BidirectionalSequenceLSTM.cpp b/nn/common/operations/BidirectionalSequenceLSTM.cpp
index 78c1a219..404179aa 100644
--- a/nn/common/operations/BidirectionalSequenceLSTM.cpp
+++ b/nn/common/operations/BidirectionalSequenceLSTM.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "BidirectionalSequenceLSTM.h"
 
 #include "CpuExecutor.h"
@@ -458,3 +459,4 @@ bool BidirectionalSequenceLSTM::Eval() {
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/BidirectionalSequenceLSTM.h b/nn/common/operations/BidirectionalSequenceLSTM.h
index 3987c230..71328530 100644
--- a/nn/common/operations/BidirectionalSequenceLSTM.h
+++ b/nn/common/operations/BidirectionalSequenceLSTM.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_BIDIRECTIONAL_SEQUENCE_LSTM_H
 #define FRAMEWORKS_ML_NN_BIDIRECTIONAL_SEQUENCE_LSTM_H
 
@@ -21,7 +22,9 @@
 #include "HalOperation.h"
 #include "LSTM.h"
 #include "OperationsUtils.h"
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/tensor_utils.h"
+#endif
 
 #include <algorithm>
 #include <cmath>
@@ -232,3 +235,4 @@ class BidirectionalSequenceLSTM {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_BIDIRECTIONAL_SEQUENCE_LSTM_H
+#endif
diff --git a/nn/common/operations/BidirectionalSequenceRNN.cpp b/nn/common/operations/BidirectionalSequenceRNN.cpp
index 32ab00fa..9a048b8a 100644
--- a/nn/common/operations/BidirectionalSequenceRNN.cpp
+++ b/nn/common/operations/BidirectionalSequenceRNN.cpp
@@ -47,6 +47,7 @@ constexpr uint32_t kMergeOutputsParam = 14;
 constexpr uint32_t kFwOutputTensor = 0;
 constexpr uint32_t kBwOutputTensor = 1;  // Only if mergeOutputs parameter is false
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -245,6 +246,7 @@ bool executeTyped(IOperationExecutionContext* context) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -269,6 +271,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     int32_t mergeOutputs = context->getInputValue<bool>(kMergeOutputsParam);
     if (mergeOutputs) {
@@ -383,6 +386,7 @@ bool execute(IOperationExecutionContext* context) {
     }
     return true;
 }
+#endif
 
 }  // namespace bidirectional_sequence_rnn
 
diff --git a/nn/common/operations/Broadcast.cpp b/nn/common/operations/Broadcast.cpp
index 25706720..5b0b943a 100644
--- a/nn/common/operations/Broadcast.cpp
+++ b/nn/common/operations/Broadcast.cpp
@@ -21,8 +21,10 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
 #include "tensorflow/lite/kernels/internal/reference/legacy_reference_ops.h"
+#endif
 
 #include "Tracing.h"
 
@@ -40,6 +42,7 @@ constexpr uint32_t kActivationScalar = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 #define ANDROID_NN_MACRO_DISPATCH(macro)                                \
@@ -346,6 +349,7 @@ bool divFloat16(const _Float16* in1, const Shape& shape1, const _Float16* in2, c
 }
 
 }  // namespace
+#endif
 
 bool validate(OperationType opType, const IOperationValidationContext* context) {
     const HalVersion opIntroducedAt = (opType == OperationType::DIV || opType == OperationType::SUB)
@@ -379,6 +383,7 @@ bool validate(OperationType opType, const IOperationValidationContext* context)
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input1 = context->getInputShape(kInputTensor1);
     Shape input2 = context->getInputShape(kInputTensor2);
@@ -512,6 +517,7 @@ bool executeDiv(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation DIV";
     }
 }
+#endif
 
 }  // namespace broadcast
 
diff --git a/nn/common/operations/Cast.cpp b/nn/common/operations/Cast.cpp
index f569767f..fee253a5 100644
--- a/nn/common/operations/Cast.cpp
+++ b/nn/common/operations/Cast.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "Cast.h"
@@ -96,3 +97,4 @@ bool eval(const uint8_t* inputData, const Shape& inputShape, uint8_t* outputData
 }  // namespace cast
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Cast.h b/nn/common/operations/Cast.h
index 47fe87ce..fb0a8c2a 100644
--- a/nn/common/operations/Cast.h
+++ b/nn/common/operations/Cast.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_CAST_H
 #define FRAMEWORKS_ML_NN_CAST_H
 
@@ -33,3 +34,4 @@ bool eval(const uint8_t* inputData, const Shape& inputShape, uint8_t* outputData
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_CAST_H
+#endif
diff --git a/nn/common/operations/ChannelShuffle.cpp b/nn/common/operations/ChannelShuffle.cpp
index 997b033f..abba1311 100644
--- a/nn/common/operations/ChannelShuffle.cpp
+++ b/nn/common/operations/ChannelShuffle.cpp
@@ -35,6 +35,7 @@ constexpr uint32_t kInputAxis = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 template <typename T>
 inline bool eval(const T* inputData, const Shape& inputShape, int32_t numGroups, int32_t axis,
                  T* outputData) {
@@ -57,6 +58,7 @@ inline bool eval(const T* inputData, const Shape& inputShape, int32_t numGroups,
     }
     return true;
 }
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -71,6 +73,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     int32_t numGroups = context->getInputValue<int32_t>(kNumGroups);
@@ -102,6 +105,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace channel_shuffle
 
diff --git a/nn/common/operations/Comparisons.cpp b/nn/common/operations/Comparisons.cpp
index 1820dde8..5ac15a8c 100644
--- a/nn/common/operations/Comparisons.cpp
+++ b/nn/common/operations/Comparisons.cpp
@@ -32,6 +32,7 @@ constexpr uint32_t kInputTensor2 = 1;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename DataType, typename ComparisonType>
@@ -119,6 +120,7 @@ bool executeGreaterTyped(IOperationExecutionContext* context) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -134,6 +136,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input1 = context->getInputShape(kInputTensor1);
     Shape input2 = context->getInputShape(kInputTensor2);
@@ -243,6 +246,7 @@ bool executeGreater(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for comparison";
     }
 }
+#endif
 
 }  // namespace comparisons
 
diff --git a/nn/common/operations/Concatenation.cpp b/nn/common/operations/Concatenation.cpp
index c0d70347..dd0a8f31 100644
--- a/nn/common/operations/Concatenation.cpp
+++ b/nn/common/operations/Concatenation.cpp
@@ -17,8 +17,10 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
 #include "tensorflow/lite/kernels/internal/reference/legacy_reference_ops.h"
+#endif
 
 #include "Tracing.h"
 
@@ -31,6 +33,7 @@ constexpr char kOperationName[] = "CONCATENATION";
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -96,6 +99,7 @@ inline bool concatenation(IOperationExecutionContext* context) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     uint32_t inputCount = context->getNumInputs();
@@ -124,6 +128,7 @@ bool validate(const IOperationValidationContext* context) {
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     uint32_t numInputs = context->getNumInputs();
     NN_RET_CHECK_GE(numInputs, 2);
@@ -168,6 +173,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace concatenation
 
diff --git a/nn/common/operations/Conv2D.cpp b/nn/common/operations/Conv2D.cpp
index ccca129f..28d47b43 100644
--- a/nn/common/operations/Conv2D.cpp
+++ b/nn/common/operations/Conv2D.cpp
@@ -19,7 +19,9 @@
 #include "Operations.h"
 
 #include "Utils.h"
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
+#endif
 
 #include "Tracing.h"
 #include <mutex>
@@ -37,6 +39,7 @@ constexpr uint32_t kBiasTensor = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 // If possible we will use this static buffer for the tensor.
@@ -401,6 +404,7 @@ bool convQuant8PerChannel(const uint8_t* inputData, const Shape& inputShape,
 #undef ANDROID_NN_CONV_PARAMETERS
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumOutputs(), kNumOutputs);
@@ -491,6 +495,7 @@ bool validate(const IOperationValidationContext* context) {
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape filter = context->getInputShape(kFilterTensor);
@@ -617,6 +622,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace conv_2d
 
diff --git a/nn/common/operations/DepthwiseConv2D.cpp b/nn/common/operations/DepthwiseConv2D.cpp
index e75d192f..15d74cd2 100644
--- a/nn/common/operations/DepthwiseConv2D.cpp
+++ b/nn/common/operations/DepthwiseConv2D.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "CpuOperationUtils.h"
 #include "Operations.h"
 
@@ -245,3 +246,4 @@ bool depthwiseConvQuant8PerChannel(const uint8_t* inputData, const Shape& inputS
 #undef ANDROID_NN_DEPTHWISE_CONV_PARAMETERS
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Dequantize.cpp b/nn/common/operations/Dequantize.cpp
index 51403d2a..1f0567b6 100644
--- a/nn/common/operations/Dequantize.cpp
+++ b/nn/common/operations/Dequantize.cpp
@@ -31,6 +31,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename InputType, typename OutputType>
@@ -73,6 +74,7 @@ bool computePerChannel(const int8_t* inputData, const Shape& inputShape, OutputT
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -96,6 +98,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     const Shape& input = context->getInputShape(kInputTensor);
     Shape output = context->getOutputShape(kOutputTensor);
@@ -140,6 +143,7 @@ bool execute(IOperationExecutionContext* context) {
     NN_RET_CHECK_FAIL() << "Unsupported tensor types combination for dequantize op. (input type: "
                         << toString(inputType) << " output type: " << toString(outputType) << ")";
 }
+#endif
 
 }  // namespace dequantize
 
diff --git a/nn/common/operations/Elementwise.cpp b/nn/common/operations/Elementwise.cpp
index 114007df..174a632e 100644
--- a/nn/common/operations/Elementwise.cpp
+++ b/nn/common/operations/Elementwise.cpp
@@ -33,6 +33,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -60,6 +61,7 @@ bool execute(IOperationExecutionContext* context, float func(float)) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -73,6 +75,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape output = context->getOutputShape(kOutputTensor);
@@ -103,6 +106,7 @@ bool executeSin(IOperationExecutionContext* context) {
 bool executeSqrt(IOperationExecutionContext* context) {
     return execute(context, std::sqrt);
 }
+#endif
 
 }  // namespace elementwise
 
diff --git a/nn/common/operations/EmbeddingLookup.cpp b/nn/common/operations/EmbeddingLookup.cpp
index c03c3435..c7b6a793 100644
--- a/nn/common/operations/EmbeddingLookup.cpp
+++ b/nn/common/operations/EmbeddingLookup.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "EmbeddingLookup.h"
 
 #include "CpuExecutor.h"
@@ -54,3 +55,4 @@ bool EmbeddingLookup::Eval() {
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/EmbeddingLookup.h b/nn/common/operations/EmbeddingLookup.h
index 9ef68227..ccdd34d6 100644
--- a/nn/common/operations/EmbeddingLookup.h
+++ b/nn/common/operations/EmbeddingLookup.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_EMBEDDING_LOOKUP_H
 #define FRAMEWORKS_ML_NN_EMBEDDING_LOOKUP_H
 
@@ -50,3 +51,4 @@ class EmbeddingLookup {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_EMBEDDING_LOOKUP_H
+#endif
diff --git a/nn/common/operations/EmbeddingLookupTest.cpp b/nn/common/operations/EmbeddingLookupTest.cpp
index f9cdf732..2add3887 100644
--- a/nn/common/operations/EmbeddingLookupTest.cpp
+++ b/nn/common/operations/EmbeddingLookupTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "EmbeddingLookup.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -172,3 +173,4 @@ TEST(EmbeddingLookupOpTest, SimpleTest) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/ExpandDims.cpp b/nn/common/operations/ExpandDims.cpp
index 2f546c98..d7cc5a43 100644
--- a/nn/common/operations/ExpandDims.cpp
+++ b/nn/common/operations/ExpandDims.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "ExpandDims.h"
@@ -47,3 +48,4 @@ bool eval(const uint8_t* inputData, const Shape& inputShape, int32_t axis, uint8
 }  // namespace expand_dims
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/ExpandDims.h b/nn/common/operations/ExpandDims.h
index 9200ca99..90ed79f5 100644
--- a/nn/common/operations/ExpandDims.h
+++ b/nn/common/operations/ExpandDims.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_EXPAND_DIMS_H
 #define FRAMEWORKS_ML_NN_EXPAND_DIMS_H
 
@@ -33,3 +34,4 @@ bool eval(const uint8_t* inputData, const Shape& inputShape, int32_t axis, uint8
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_EXPAND_DIMS_H
+#endif
diff --git a/nn/common/operations/FullyConnected.cpp b/nn/common/operations/FullyConnected.cpp
index 93667a45..c75267db 100644
--- a/nn/common/operations/FullyConnected.cpp
+++ b/nn/common/operations/FullyConnected.cpp
@@ -17,8 +17,10 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
 #include "tensorflow/lite/kernels/internal/reference/reference_ops.h"
+#endif
 
 #include "Tracing.h"
 #include <mutex>
@@ -38,6 +40,7 @@ constexpr uint32_t kActivationScalar = 3;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 // executionMutex is used to protect concurrent access of non-threadsafe resources
@@ -138,6 +141,7 @@ bool fullyConnectedQuant8(const uint8_t* inputData, const Shape& inputShape,
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -191,6 +195,7 @@ bool validate(const IOperationValidationContext* context) {
     return true;
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape weights = context->getInputShape(kWeightsTensor);
@@ -262,6 +267,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace fully_connected
 
diff --git a/nn/common/operations/Gather.cpp b/nn/common/operations/Gather.cpp
index 0dc976ab..3f62f029 100644
--- a/nn/common/operations/Gather.cpp
+++ b/nn/common/operations/Gather.cpp
@@ -36,6 +36,7 @@ constexpr uint32_t kInputIndices = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -60,6 +61,7 @@ inline bool eval(const T* inputData, const Shape& inputShape, int32_t axis,
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -75,6 +77,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     int32_t axis = context->getInputValue<int32_t>(kInputAxis);
@@ -126,6 +129,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace gather
 
diff --git a/nn/common/operations/GenerateProposals.cpp b/nn/common/operations/GenerateProposals.cpp
index 9b0df049..1d4f8078 100644
--- a/nn/common/operations/GenerateProposals.cpp
+++ b/nn/common/operations/GenerateProposals.cpp
@@ -28,6 +28,7 @@ namespace android {
 namespace nn {
 namespace bbox_ops {
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 struct BoxEncodingCorner {
@@ -157,6 +158,7 @@ float getIoUAxisAligned(const float* roi1, const float* roi2) {
 }
 
 }  // namespace
+#endif
 
 namespace axis_aligned_bbox_transform {
 
@@ -190,6 +192,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape roiShape = context->getInputShape(kRoiTensor);
     Shape bboxDeltasShape = context->getInputShape(kDeltaTensor);
@@ -275,6 +278,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace axis_aligned_bbox_transform
 
@@ -299,6 +303,7 @@ constexpr uint32_t kOutputRoiTensor = 1;
 constexpr uint32_t kOutputClassTensor = 2;
 constexpr uint32_t kOutputBatchesTensor = 3;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 // TODO(xusongw): Reduce code duplication with hard/soft nms path.
@@ -628,6 +633,7 @@ bool boxWithNmsLimitQuant(const uint8_t* scoresData, const Shape& scoresShape,
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -669,6 +675,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape scoreShape = context->getInputShape(kScoreTensor);
     Shape roiShape = context->getInputShape(kRoiTensor);
@@ -794,6 +801,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace box_with_nms_limit
 
@@ -819,6 +827,7 @@ constexpr uint32_t kOutputScoreTensor = 0;
 constexpr uint32_t kOutputRoiTensor = 1;
 constexpr uint32_t kOutputBatchesTensor = 2;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 void filterBoxes(const float* roiBase, const float* imageInfoBase, float minSize,
@@ -1104,6 +1113,7 @@ bool generateProposalsQuant(const uint8_t* scoresData, const Shape& scoresShape,
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -1161,6 +1171,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     bool useNchw = context->getInputValue<bool>(kLayoutScalar);
     Shape scoreShape = context->getInputShape(kScoreTensor);
@@ -1273,6 +1284,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace generate_proposals
 
@@ -1302,6 +1314,7 @@ constexpr uint32_t kOutputRoiTensor = 1;
 constexpr uint32_t kOutputClassTensor = 2;
 constexpr uint32_t kOutputDetectionTensor = 3;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 bool detectionPostprocessFloat32(
@@ -1438,6 +1451,7 @@ bool detectionPostprocessFloat16(
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -1470,6 +1484,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape scoreShape = context->getInputShape(kScoreTensor);
     Shape deltasShape = context->getInputShape(kDeltaTensor);
@@ -1606,6 +1621,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace detection_postprocess
 
diff --git a/nn/common/operations/GroupedConv2D.cpp b/nn/common/operations/GroupedConv2D.cpp
index 2644f80a..5353f536 100644
--- a/nn/common/operations/GroupedConv2D.cpp
+++ b/nn/common/operations/GroupedConv2D.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "CpuOperationUtils.h"
 #include "Operations.h"
 
@@ -288,3 +289,4 @@ bool groupedConvFloat16(const _Float16* inputData, const Shape& inputShape,
 #undef ANDROID_NN_GROUPED_CONV_PARAMETERS
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/HalOperation.h b/nn/common/operations/HalOperation.h
index 001e9372..5449d938 100644
--- a/nn/common/operations/HalOperation.h
+++ b/nn/common/operations/HalOperation.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_HAL_OPERATION_H
 #define FRAMEWORKS_ML_NN_HAL_OPERATION_H
 
@@ -35,3 +36,4 @@ struct Operation;
 using ::android::hardware::neuralnetworks::V1_2::Operation;
 
 #endif  // FRAMEWORKS_ML_NN_HAL_OPERATION_H
+#endif
diff --git a/nn/common/operations/HashtableLookup.cpp b/nn/common/operations/HashtableLookup.cpp
index f8ee5f63..40d7f2af 100644
--- a/nn/common/operations/HashtableLookup.cpp
+++ b/nn/common/operations/HashtableLookup.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "HashtableLookup.h"
 
 #include "CpuExecutor.h"
@@ -72,3 +73,4 @@ bool HashtableLookup::Eval() {
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/HashtableLookup.h b/nn/common/operations/HashtableLookup.h
index 3cf28aa2..fdd326ef 100644
--- a/nn/common/operations/HashtableLookup.h
+++ b/nn/common/operations/HashtableLookup.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_HASHTABLE_LOOKUP_H
 #define FRAMEWORKS_ML_NN_HASHTABLE_LOOKUP_H
 
@@ -54,3 +55,4 @@ class HashtableLookup {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_HASHTABLE_LOOKUP_H
+#endif
diff --git a/nn/common/operations/HashtableLookupTest.cpp b/nn/common/operations/HashtableLookupTest.cpp
index 103f5996..5fc4ec80 100644
--- a/nn/common/operations/HashtableLookupTest.cpp
+++ b/nn/common/operations/HashtableLookupTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "HashtableLookup.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -187,3 +188,4 @@ TEST(HashtableLookupOpTest, BlackBoxTest) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/HeatmapMaxKeypoint.cpp b/nn/common/operations/HeatmapMaxKeypoint.cpp
index 0e3e9761..ca52e130 100644
--- a/nn/common/operations/HeatmapMaxKeypoint.cpp
+++ b/nn/common/operations/HeatmapMaxKeypoint.cpp
@@ -38,6 +38,7 @@ constexpr uint32_t kNumOutputs = 2;
 constexpr uint32_t kOutputScoreTensor = 0;
 constexpr uint32_t kOutputKeypointTensor = 1;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 // This function uses Taylor expansion up to the quatratic term to approximate bicubic
@@ -198,6 +199,7 @@ inline bool heatmapMaxKeypointQuant(const uint8_t* heatmap, const Shape& heatmap
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -221,6 +223,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     bool layout = context->getInputValue<bool>(kLayoutScalar);
     Shape heatmapShape = context->getInputShape(kHeatmapTensor);
@@ -308,6 +311,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace heatmap_max_keypoint
 
diff --git a/nn/common/operations/InstanceNormalization.cpp b/nn/common/operations/InstanceNormalization.cpp
index c61da096..1fe7db0a 100644
--- a/nn/common/operations/InstanceNormalization.cpp
+++ b/nn/common/operations/InstanceNormalization.cpp
@@ -40,6 +40,7 @@ constexpr uint32_t kLayoutScalar = 4;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -88,6 +89,7 @@ inline bool instanceNorm(const T* inputData, const Shape& inputShape, T gamma, T
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -109,6 +111,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     NN_RET_CHECK_EQ(getNumberOfDimensions(input), 4);
@@ -139,6 +142,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace instance_normalization
 
diff --git a/nn/common/operations/L2Normalization.cpp b/nn/common/operations/L2Normalization.cpp
index b122410d..9765f1c2 100644
--- a/nn/common/operations/L2Normalization.cpp
+++ b/nn/common/operations/L2Normalization.cpp
@@ -17,7 +17,9 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/optimized_ops.h"
+#endif
 
 #include "Tracing.h"
 
@@ -34,6 +36,7 @@ constexpr uint32_t kAxisScalar = 1;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 inline bool l2normFloat32Impl(const float* inputData, const Shape& inputShape, int32_t axis,
@@ -141,6 +144,7 @@ bool l2normQuant8(const uint8_t* inputData, const Shape& inputShape, int32_t axi
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK(context->getNumInputs() == kNumInputs ||
@@ -166,6 +170,7 @@ bool validate(const IOperationValidationContext* context) {
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     const Shape& input = context->getInputShape(kInputTensor);
     int32_t numDimensions = getNumberOfDimensions(input);
@@ -212,6 +217,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace l2_norm
 
diff --git a/nn/common/operations/LSHProjection.cpp b/nn/common/operations/LSHProjection.cpp
index 353821a9..522e733f 100644
--- a/nn/common/operations/LSHProjection.cpp
+++ b/nn/common/operations/LSHProjection.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "LSHProjection.h"
 
 #include "CpuExecutor.h"
@@ -191,3 +192,4 @@ template void DenseLshProjection<_Float16>(const RunTimeOperandInfo* hash,
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/LSHProjection.h b/nn/common/operations/LSHProjection.h
index ec157e31..196af59b 100644
--- a/nn/common/operations/LSHProjection.h
+++ b/nn/common/operations/LSHProjection.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_LSH_PROJECTION_H
 #define FRAMEWORKS_ML_NN_LSH_PROJECTION_H
 
@@ -78,3 +79,4 @@ void DenseLshProjection(const RunTimeOperandInfo* hash, const RunTimeOperandInfo
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_LSH_PROJECTION_H
+#endif
diff --git a/nn/common/operations/LSHProjectionTest.cpp b/nn/common/operations/LSHProjectionTest.cpp
index bdbda55b..9ac58153 100644
--- a/nn/common/operations/LSHProjectionTest.cpp
+++ b/nn/common/operations/LSHProjectionTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "LSHProjection.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -172,3 +173,4 @@ TEST(LSHProjectionOpTest2, SparseWithTwoInputs) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/LSTM.cpp b/nn/common/operations/LSTM.cpp
index b219e753..18f99690 100644
--- a/nn/common/operations/LSTM.cpp
+++ b/nn/common/operations/LSTM.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "LSTM.h"
 
 #include "CpuExecutor.h"
@@ -1032,3 +1033,4 @@ bool LSTMCell::Eval() {
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/LSTM.h b/nn/common/operations/LSTM.h
index 86fe63df..f8eca85b 100644
--- a/nn/common/operations/LSTM.h
+++ b/nn/common/operations/LSTM.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_LSTMCELL_H
 #define FRAMEWORKS_ML_NN_LSTMCELL_H
 
@@ -248,3 +249,4 @@ class LSTMCell {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_LSTMCELL_H
+#endif
diff --git a/nn/common/operations/LSTMTest.cpp b/nn/common/operations/LSTMTest.cpp
index b990b8b9..01630ff2 100644
--- a/nn/common/operations/LSTMTest.cpp
+++ b/nn/common/operations/LSTMTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "LSTM.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -1111,3 +1112,4 @@ TEST(LSTMOpTest, BlackBoxTestWithPeepholeWithProjectionNoClipping) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/LayerNormLSTMTest.cpp b/nn/common/operations/LayerNormLSTMTest.cpp
index faf7fef1..87c5b0fb 100644
--- a/nn/common/operations/LayerNormLSTMTest.cpp
+++ b/nn/common/operations/LayerNormLSTMTest.cpp
@@ -13,6 +13,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "LSTM.h"
 
 #include <android-base/logging.h>
@@ -426,3 +428,4 @@ TEST(LSTMOpTest, LayerNormNoCifgPeepholeProjectionNoClipping) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/LogSoftmax.cpp b/nn/common/operations/LogSoftmax.cpp
index d0a5cb9a..362dd544 100644
--- a/nn/common/operations/LogSoftmax.cpp
+++ b/nn/common/operations/LogSoftmax.cpp
@@ -37,6 +37,7 @@ constexpr uint32_t kInputAxis = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 template <typename T>
 inline bool compute(const T* input, const Shape& shape, T beta, uint32_t axis, T* output) {
     const uint32_t outerSize = getNumberOfElements(shape, 0, axis);
@@ -68,6 +69,7 @@ inline bool compute(const T* input, const Shape& shape, T beta, uint32_t axis, T
     }
     return true;
 }
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -90,6 +92,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     return context->setOutputShape(kOutputTensor, context->getInputShape(kInputTensor));
 }
@@ -112,6 +115,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace log_softmax
 
diff --git a/nn/common/operations/LogicalAndOr.cpp b/nn/common/operations/LogicalAndOr.cpp
index 9d4968df..1728a546 100644
--- a/nn/common/operations/LogicalAndOr.cpp
+++ b/nn/common/operations/LogicalAndOr.cpp
@@ -32,6 +32,7 @@ constexpr uint32_t kInputTensor2 = 1;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 bool compute(const std::function<bool(bool, bool)>& func, const bool8* aData, const Shape& aShape,
@@ -57,6 +58,7 @@ bool compute(const std::function<bool(bool, bool)>& func, const bool8* aData, co
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -69,6 +71,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input1 = context->getInputShape(kInputTensor1);
     Shape input2 = context->getInputShape(kInputTensor2);
@@ -92,6 +95,7 @@ bool executeOr(IOperationExecutionContext* context) {
             context->getInputShape(kInputTensor2), context->getOutputBuffer<bool8>(kOutputTensor),
             context->getOutputShape(kOutputTensor));
 }
+#endif
 
 }  // namespace logical
 
diff --git a/nn/common/operations/LogicalNot.cpp b/nn/common/operations/LogicalNot.cpp
index c7153885..b7fecc80 100644
--- a/nn/common/operations/LogicalNot.cpp
+++ b/nn/common/operations/LogicalNot.cpp
@@ -29,6 +29,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 bool compute(const bool8* input, const Shape& shape, bool8* output) {
@@ -40,6 +41,7 @@ bool compute(const bool8* input, const Shape& shape, bool8* output) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -52,6 +54,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape output = context->getOutputShape(kOutputTensor);
@@ -64,6 +67,7 @@ bool execute(IOperationExecutionContext* context) {
                    context->getInputShape(kInputTensor),
                    context->getOutputBuffer<bool8>(kOutputTensor));
 }
+#endif
 
 }  // namespace logical_not
 
diff --git a/nn/common/operations/MaximumMinimum.cpp b/nn/common/operations/MaximumMinimum.cpp
index d56e3118..b0a02e1f 100644
--- a/nn/common/operations/MaximumMinimum.cpp
+++ b/nn/common/operations/MaximumMinimum.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "MaximumMinimum.h"
@@ -122,3 +123,4 @@ bool eval(const void* in1, const Shape& shape1, const void* in2, const Shape& sh
 }  // namespace maximum_minimum
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/MaximumMinimum.h b/nn/common/operations/MaximumMinimum.h
index 8ef7ccdb..fc293c7a 100644
--- a/nn/common/operations/MaximumMinimum.h
+++ b/nn/common/operations/MaximumMinimum.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_MAXIMUM_MINIMUM_H
 #define FRAMEWORKS_ML_NN_MAXIMUM_MINIMUM_H
 
@@ -33,3 +34,4 @@ bool eval(const void* in1, const Shape& shape1, const void* in2, const Shape& sh
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_MAXIMUM_MINIMUM_H
+#endif
diff --git a/nn/common/operations/Multinomial.cpp b/nn/common/operations/Multinomial.cpp
index 4fcd482d..f8c00805 100644
--- a/nn/common/operations/Multinomial.cpp
+++ b/nn/common/operations/Multinomial.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "Multinomial.h"
 
 #include "CpuExecutor.h"
@@ -141,3 +142,4 @@ void Multinomial::EvalFloat32(const float* inputData) {
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Multinomial.h b/nn/common/operations/Multinomial.h
index 8fc3b104..3bffd6c9 100644
--- a/nn/common/operations/Multinomial.h
+++ b/nn/common/operations/Multinomial.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_MULTINOMIAL_H
 #define FRAMEWORKS_ML_NN_MULTINOMIAL_H
 
@@ -60,3 +61,4 @@ class Multinomial {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_MULTINOMIAL_H
+#endif
diff --git a/nn/common/operations/MultinomialTest.cpp b/nn/common/operations/MultinomialTest.cpp
index 4cc66f49..6527a684 100644
--- a/nn/common/operations/MultinomialTest.cpp
+++ b/nn/common/operations/MultinomialTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "Multinomial.h"
 
 #include "HalInterfaces.h"
@@ -137,3 +138,4 @@ TEST(MultinomialOpTest, ProbabilityDeltaWithinTolerance) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Neg.cpp b/nn/common/operations/Neg.cpp
index 7c610284..97029f4a 100644
--- a/nn/common/operations/Neg.cpp
+++ b/nn/common/operations/Neg.cpp
@@ -35,6 +35,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -47,6 +48,7 @@ inline bool compute(const T* input, const Shape& shape, T* output) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -60,6 +62,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape output = context->getOutputShape(kOutputTensor);
@@ -85,6 +88,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace neg
 
diff --git a/nn/common/operations/Normalization.cpp b/nn/common/operations/Normalization.cpp
index 89554b0a..a7424992 100644
--- a/nn/common/operations/Normalization.cpp
+++ b/nn/common/operations/Normalization.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "CpuOperationUtils.h"
 #include "Operations.h"
 
@@ -92,3 +93,4 @@ bool localResponseNormFloat32(const float* inputData, const Shape& inputShape, i
 }
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/PRelu.cpp b/nn/common/operations/PRelu.cpp
index 68d1e5bc..03c9fca0 100644
--- a/nn/common/operations/PRelu.cpp
+++ b/nn/common/operations/PRelu.cpp
@@ -21,7 +21,9 @@
 #include "OperationResolver.h"
 #include "OperationsUtils.h"
 #include "Tracing.h"
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
+#endif
 
 namespace android {
 namespace nn {
@@ -36,6 +38,7 @@ constexpr uint32_t kAlphaTensor = 1;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 template <typename T>
 inline bool eval(const std::function<T(const T&, const T&)>& func, const T* aData,
                  const Shape& aShape, const T* bData, const Shape& bShape, T* outputData,
@@ -91,6 +94,7 @@ bool evalQuant8(const uint8_t* aData, const Shape& aShape, const uint8_t* bData,
             },
             aData, aShape, bData, bShape, outputData, outputShape);
 }
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -105,6 +109,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape alpha = context->getInputShape(kAlphaTensor);
@@ -150,6 +155,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace prelu
 
diff --git a/nn/common/operations/Pooling.cpp b/nn/common/operations/Pooling.cpp
index c32617f2..f12222c7 100644
--- a/nn/common/operations/Pooling.cpp
+++ b/nn/common/operations/Pooling.cpp
@@ -17,7 +17,9 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/optimized_ops.h"
+#endif
 
 #include "Tracing.h"
 
@@ -30,6 +32,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 struct PoolingParam {
@@ -246,6 +249,7 @@ bool maxPool(const T* inputData, const Shape& inputShape, const PoolingParam& pa
 }
 
 }  // namespace
+#endif
 
 bool validate(OperationType opType, const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumOutputs(), kNumOutputs);
@@ -298,6 +302,7 @@ bool validate(OperationType opType, const IOperationValidationContext* context)
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     NN_RET_CHECK_EQ(getNumberOfDimensions(input), 4);
@@ -375,6 +380,7 @@ bool executeMaxPool(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation MAX_POOL_2D";
     }
 }
+#endif
 
 #undef POOLING_DISPATCH_INPUT_TYPE
 
diff --git a/nn/common/operations/Pow.cpp b/nn/common/operations/Pow.cpp
index 99e10998..f3556084 100644
--- a/nn/common/operations/Pow.cpp
+++ b/nn/common/operations/Pow.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "Pow.h"
@@ -87,3 +88,4 @@ bool eval(const void* baseData, const Shape& baseShape, const void* exponentData
 }  // namespace pow
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Pow.h b/nn/common/operations/Pow.h
index 6830992b..3d884eb3 100644
--- a/nn/common/operations/Pow.h
+++ b/nn/common/operations/Pow.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_POW_H
 #define FRAMEWORKS_ML_NN_POW_H
 
@@ -33,3 +34,4 @@ bool eval(const void* baseData, const Shape& baseShape, const void* exponentData
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_POW_H
+#endif
diff --git a/nn/common/operations/Quantize.cpp b/nn/common/operations/Quantize.cpp
index b2ec9e34..bb34f33b 100644
--- a/nn/common/operations/Quantize.cpp
+++ b/nn/common/operations/Quantize.cpp
@@ -34,6 +34,7 @@ constexpr uint32_t kInputTensor = 0;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 bool quantizeFloat32ToQuant8(const float* inputData, uint8_t* outputData,
@@ -61,6 +62,7 @@ bool quantizeFloat16ToQuant8(const _Float16* inputData, uint8_t* outputData,
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -77,6 +79,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     const Shape& input = context->getInputShape(kInputTensor);
     Shape output = context->getOutputShape(kOutputTensor);
@@ -103,6 +106,7 @@ bool execute(IOperationExecutionContext* context) {
                         << " output type: " << toString(context->getOutputType(kOutputTensor))
                         << ")";
 }
+#endif
 
 }  // namespace quantize
 
diff --git a/nn/common/operations/QuantizedLSTM.cpp b/nn/common/operations/QuantizedLSTM.cpp
index 538cd19e..e48415b4 100644
--- a/nn/common/operations/QuantizedLSTM.cpp
+++ b/nn/common/operations/QuantizedLSTM.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "QuantizedLSTM.h"
 
 #include "CpuExecutor.h"
@@ -423,3 +424,4 @@ bool QuantizedLSTMCell::eval() {
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/QuantizedLSTM.h b/nn/common/operations/QuantizedLSTM.h
index 11dc02b5..83b6cb81 100644
--- a/nn/common/operations/QuantizedLSTM.h
+++ b/nn/common/operations/QuantizedLSTM.h
@@ -1,3 +1,4 @@
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_QUANTIZEDLSTM_H
 #define FRAMEWORKS_ML_NN_QUANTIZEDLSTM_H
 
@@ -80,3 +81,4 @@ class QuantizedLSTMCell {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_QUANTIZEDLSTM_H
+#endif
diff --git a/nn/common/operations/QuantizedLSTMTest.cpp b/nn/common/operations/QuantizedLSTMTest.cpp
index 395f740c..ae560ef9 100644
--- a/nn/common/operations/QuantizedLSTMTest.cpp
+++ b/nn/common/operations/QuantizedLSTMTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "QuantizedLSTM.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -375,3 +376,4 @@ TEST_F(QuantizedLstmTest, BasicQuantizedLstmTest) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/RNN.cpp b/nn/common/operations/RNN.cpp
index e91640e5..c2938f92 100644
--- a/nn/common/operations/RNN.cpp
+++ b/nn/common/operations/RNN.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "RNN.h"
 
 #include "CpuExecutor.h"
@@ -252,3 +253,4 @@ template bool RNN::RNNStep<_Float16>(const _Float16* inputData, const Shape& inp
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/RNN.h b/nn/common/operations/RNN.h
index ae7cf899..bb5e247e 100644
--- a/nn/common/operations/RNN.h
+++ b/nn/common/operations/RNN.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_RNN_H
 #define FRAMEWORKS_ML_NN_RNN_H
 
@@ -77,3 +78,4 @@ class RNN {
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_RNN_H
+#endif
diff --git a/nn/common/operations/RNNTest.cpp b/nn/common/operations/RNNTest.cpp
index b5833e2b..8ff12a90 100644
--- a/nn/common/operations/RNNTest.cpp
+++ b/nn/common/operations/RNNTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "RNN.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -335,3 +336,4 @@ TEST(RNNOpTest, BlackBoxTest) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Reduce.cpp b/nn/common/operations/Reduce.cpp
index 3f870522..d2dda642 100644
--- a/nn/common/operations/Reduce.cpp
+++ b/nn/common/operations/Reduce.cpp
@@ -40,6 +40,7 @@ constexpr uint32_t kOutputTensor = 0;
 constexpr _Float16 kFloat16Max = 65504;
 constexpr _Float16 kFloat16Lowest = -kFloat16Max;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -62,6 +63,7 @@ inline bool compute(IOperationExecutionContext* context, T init, T func(T, T)) {
 }
 
 }  // namespace
+#endif
 
 bool validateProdSum(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -102,6 +104,7 @@ bool validateLogical(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape inputShape = context->getInputShape(kInputTensor);
     const uint32_t inputRank = getNumberOfDimensions(inputShape);
@@ -207,6 +210,7 @@ bool executeAll(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation REDUCE_ALL";
     }
 }
+#endif
 
 }  // namespace reduce
 
diff --git a/nn/common/operations/Reshape.cpp b/nn/common/operations/Reshape.cpp
index 2633812c..4cb14780 100644
--- a/nn/common/operations/Reshape.cpp
+++ b/nn/common/operations/Reshape.cpp
@@ -16,6 +16,7 @@
 
 // Contains the implementation of the operations.
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "CpuOperationUtils.h"
@@ -233,3 +234,4 @@ template bool spaceToBatchGeneric<uint8_t>(const uint8_t* inputData, const Shape
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/ResizeImageOps.cpp b/nn/common/operations/ResizeImageOps.cpp
index 57a94f6c..61f88948 100644
--- a/nn/common/operations/ResizeImageOps.cpp
+++ b/nn/common/operations/ResizeImageOps.cpp
@@ -21,7 +21,9 @@
 #include "OperationResolver.h"
 #include "Tracing.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/reference/reference_ops.h"
+#endif
 
 #include <functional>
 #include <vector>
@@ -42,6 +44,7 @@ constexpr uint32_t kLayoutScalar = 4;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -100,6 +103,7 @@ bool resizeImageOp(OperationType opType, const T* inputData, const Shape& inputS
 }
 
 }  // namespace
+#endif
 
 bool validate(OperationType opType, const IOperationValidationContext* context) {
     if (opType == OperationType::RESIZE_BILINEAR) {
@@ -142,6 +146,7 @@ bool validate(OperationType opType, const IOperationValidationContext* context)
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(OperationType opType, IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     NN_RET_CHECK_EQ(getNumberOfDimensions(input), 4);
@@ -224,6 +229,7 @@ bool execute(OperationType opType, IOperationExecutionContext* context) {
                                 << getOperationName(opType);
     }
 }
+#endif
 
 }  // namespace resize_image
 
diff --git a/nn/common/operations/RoiAlign.cpp b/nn/common/operations/RoiAlign.cpp
index 578c5567..193c8a89 100644
--- a/nn/common/operations/RoiAlign.cpp
+++ b/nn/common/operations/RoiAlign.cpp
@@ -22,7 +22,9 @@
 #include <cmath>
 
 #include "Tracing.h"
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/common.h"
+#endif
 
 namespace android {
 namespace nn {
@@ -45,6 +47,7 @@ constexpr uint32_t kLayoutScalar = 9;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T_Input, typename T_Roi>
@@ -326,6 +329,7 @@ inline bool roiAlign(const T_Input* inputData, const Shape& inputShape, const T_
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -364,6 +368,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     bool useNchw = context->getInputValue<bool>(kLayoutScalar);
     Shape input = context->getInputShape(kInputTensor);
@@ -474,6 +479,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace roi_align
 
diff --git a/nn/common/operations/RoiPooling.cpp b/nn/common/operations/RoiPooling.cpp
index 37914fdc..6920bae7 100644
--- a/nn/common/operations/RoiPooling.cpp
+++ b/nn/common/operations/RoiPooling.cpp
@@ -42,6 +42,7 @@ constexpr uint32_t kLayoutScalar = 7;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T_Input, typename T_Roi>
@@ -165,6 +166,7 @@ inline bool roiPooling<uint8_t, uint16_t>(const uint8_t* inputData, const Shape&
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -199,6 +201,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     bool useNchw = context->getInputValue<bool>(kLayoutScalar);
     Shape input = context->getInputShape(kInputTensor);
@@ -288,6 +291,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace roi_pooling
 
diff --git a/nn/common/operations/SVDF.cpp b/nn/common/operations/SVDF.cpp
index ac3ce357..c75c5e60 100644
--- a/nn/common/operations/SVDF.cpp
+++ b/nn/common/operations/SVDF.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "SVDF.h"
 
 #include "CpuExecutor.h"
@@ -221,3 +222,4 @@ void SVDF::EvalFloat32(const float* inputData, const float* inputStateData, cons
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/SVDF.h b/nn/common/operations/SVDF.h
index e868dc88..048a5e00 100644
--- a/nn/common/operations/SVDF.h
+++ b/nn/common/operations/SVDF.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_SVDF_H
 #define FRAMEWORKS_ML_NN_SVDF_H
 
@@ -42,7 +43,6 @@ class SVDF {
     static bool Prepare(const Operation& operation, std::vector<RunTimeOperandInfo>& operands,
                         Shape* stateShape, Shape* outputShape);
     bool Eval();
-
     static constexpr int kInputTensor = 0;
     static constexpr int kWeightsFeatureTensor = 1;
     static constexpr int kWeightsTimeTensor = 2;
@@ -75,3 +75,4 @@ class SVDF {
 }  // namespace android
 
 #endif
+#endif
diff --git a/nn/common/operations/SVDFTest.cpp b/nn/common/operations/SVDFTest.cpp
index da96dd2b..010caeea 100644
--- a/nn/common/operations/SVDFTest.cpp
+++ b/nn/common/operations/SVDFTest.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "SVDF.h"
 
 #include "NeuralNetworksWrapper.h"
@@ -447,3 +448,4 @@ TEST(SVDFOpTest, BlackBoxTestRank2) {
 }  // namespace wrapper
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Select.cpp b/nn/common/operations/Select.cpp
index 2a26c1e6..1cb42942 100644
--- a/nn/common/operations/Select.cpp
+++ b/nn/common/operations/Select.cpp
@@ -33,6 +33,7 @@ constexpr uint32_t kInputTensor2 = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -65,6 +66,7 @@ bool executeTyped(IOperationExecutionContext* context) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -79,6 +81,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape inputCondition = context->getInputShape(kInputCondition);
     Shape input1 = context->getInputShape(kInputTensor1);
@@ -115,6 +118,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for SELECT op.";
     }
 }
+#endif
 
 }  // namespace select_op
 
diff --git a/nn/common/operations/SimpleMath.cpp b/nn/common/operations/SimpleMath.cpp
index 62fbeb3d..8e4dabcd 100644
--- a/nn/common/operations/SimpleMath.cpp
+++ b/nn/common/operations/SimpleMath.cpp
@@ -16,6 +16,7 @@
 
 // Contains the implementation of the operations.
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "CpuOperationUtils.h"
@@ -102,3 +103,4 @@ template bool meanGeneric<uint8_t, int32_t>(uint8_t* inputData, const Shape& inp
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Slice.cpp b/nn/common/operations/Slice.cpp
index 67663f48..d96168fe 100644
--- a/nn/common/operations/Slice.cpp
+++ b/nn/common/operations/Slice.cpp
@@ -33,6 +33,7 @@ constexpr uint32_t kSizeTensor = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -74,6 +75,7 @@ bool evalGeneric(const T* inputData, const Shape& inputShape, const int32_t* beg
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -122,6 +124,7 @@ bool prepare(IOperationExecutionContext* context) {
     return context->setOutputShape(kOutputTensor, outputShape);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool execute(IOperationExecutionContext* context) {
     // Bypass execution in the case of zero-sized input.
     if (getNumberOfElements(context->getOutputShape(kOutputTensor)) == 0) return true;
@@ -166,6 +169,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace slice
 
diff --git a/nn/common/operations/Softmax.cpp b/nn/common/operations/Softmax.cpp
index 06c419ec..3c9b66a8 100644
--- a/nn/common/operations/Softmax.cpp
+++ b/nn/common/operations/Softmax.cpp
@@ -17,8 +17,10 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
 #include "tensorflow/lite/kernels/internal/optimized/optimized_ops.h"
+#endif
 
 #include "Tracing.h"
 #include <float.h>
@@ -38,6 +40,7 @@ constexpr uint32_t kAxisScalar = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 inline bool softmaxSlowFloat32(const float* inputData, const Shape& inputShape, const float beta,
@@ -223,6 +226,7 @@ bool softmaxQuant8(const uint8_t* inputData, const Shape& inputShape, const floa
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK(context->getNumInputs() == kNumInputs ||
@@ -252,6 +256,7 @@ bool validate(const IOperationValidationContext* context) {
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     float beta = (input.type == OperandType::TENSOR_FLOAT16)
@@ -293,6 +298,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace softmax
 
diff --git a/nn/common/operations/Split.cpp b/nn/common/operations/Split.cpp
index e37af519..9e85a9ac 100644
--- a/nn/common/operations/Split.cpp
+++ b/nn/common/operations/Split.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "Operations.h"
@@ -81,3 +82,4 @@ bool splitInt32(const int32_t* inputData, const Shape& inputShape, int32_t axis,
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/StridedSlice.cpp b/nn/common/operations/StridedSlice.cpp
index c58db7dc..26c50bb6 100644
--- a/nn/common/operations/StridedSlice.cpp
+++ b/nn/common/operations/StridedSlice.cpp
@@ -16,6 +16,7 @@
 
 // Contains the implementation of the operations.
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "CpuOperationUtils.h"
@@ -86,3 +87,4 @@ bool stridedSliceGeneric(const uint8_t* inputData, const Shape& inputShape,
 
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Tile.cpp b/nn/common/operations/Tile.cpp
index e5f4a612..c223c15a 100644
--- a/nn/common/operations/Tile.cpp
+++ b/nn/common/operations/Tile.cpp
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #define LOG_TAG "Operations"
 
 #include "Tile.h"
@@ -109,3 +110,4 @@ bool eval(const uint8_t* inputData, const Shape& inputShape, const int32_t* mult
 }  // namespace tile
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/Tile.h b/nn/common/operations/Tile.h
index 9ba70a6f..0d869138 100644
--- a/nn/common/operations/Tile.h
+++ b/nn/common/operations/Tile.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_COMMON_OPERATIONS_TILE_H
 #define FRAMEWORKS_ML_NN_COMMON_OPERATIONS_TILE_H
 
@@ -34,3 +35,4 @@ bool eval(const uint8_t* inputData, const Shape& inputShape, const int32_t* mult
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_COMMON_OPERATIONS_TILE_H
+#endif
diff --git a/nn/common/operations/TopK_V2.cpp b/nn/common/operations/TopK_V2.cpp
index 95b4ae18..dceca827 100644
--- a/nn/common/operations/TopK_V2.cpp
+++ b/nn/common/operations/TopK_V2.cpp
@@ -13,6 +13,8 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "TopK_V2.h"
 
 #include "OperationsUtils.h"
@@ -97,3 +99,4 @@ bool eval(const void* inputData, const Shape& inputShape, const int32_t k, void*
 }  // namespace topk_v2
 }  // namespace nn
 }  // namespace android
+#endif
diff --git a/nn/common/operations/TopK_V2.h b/nn/common/operations/TopK_V2.h
index daad30e3..eb2d65ea 100644
--- a/nn/common/operations/TopK_V2.h
+++ b/nn/common/operations/TopK_V2.h
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #ifndef FRAMEWORKS_ML_NN_TOPK_V2_H
 #define FRAMEWORKS_ML_NN_TOPK_V2_H
 
@@ -33,3 +34,4 @@ bool eval(const void* inputData, const Shape& inputShape, const int32_t k, void*
 }  // namespace android
 
 #endif  // FRAMEWORKS_ML_NN_TOPK_V2_H
+#endif
diff --git a/nn/common/operations/Transpose.cpp b/nn/common/operations/Transpose.cpp
index 3b417f2b..2116a8e4 100644
--- a/nn/common/operations/Transpose.cpp
+++ b/nn/common/operations/Transpose.cpp
@@ -17,8 +17,10 @@
 #include "CpuOperationUtils.h"
 #include "OperationResolver.h"
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/optimized/legacy_optimized_ops.h"
 #include "tensorflow/lite/kernels/internal/reference/reference_ops.h"
+#endif
 
 #include "Tracing.h"
 
@@ -35,6 +37,7 @@ constexpr uint32_t kPermTensor = 1;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -65,6 +68,7 @@ bool transposeGeneric(const T* inputData, const Shape& inputShape, const int32_t
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -82,6 +86,7 @@ bool validate(const IOperationValidationContext* context) {
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     // Only the permutation tensor can be omitted.
     NN_RET_CHECK(!context->isOmittedInput(kInputTensor));
@@ -150,6 +155,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace transpose
 
diff --git a/nn/common/operations/TransposeConv2D.cpp b/nn/common/operations/TransposeConv2D.cpp
index 4aa8cd5d..bf5130e4 100644
--- a/nn/common/operations/TransposeConv2D.cpp
+++ b/nn/common/operations/TransposeConv2D.cpp
@@ -21,7 +21,9 @@
 #include <cmath>
 
 #include "Tracing.h"
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 #include "tensorflow/lite/kernels/internal/common.h"
+#endif
 #include <mutex>
 
 namespace android {
@@ -37,6 +39,7 @@ constexpr uint32_t kBiasTensor = 2;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 // If possible we will use this static buffer for the tensor.
@@ -424,6 +427,7 @@ bool transposeConvQuant8PerChannel(const uint8_t* inputData, const Shape& inputS
 #undef ANDROID_NN_TRANSPOSE_CONV_PARAMETERS
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumOutputs(), kNumOutputs);
@@ -463,6 +467,7 @@ bool validate(const IOperationValidationContext* context) {
            validateOutputTypes(context, {inputType});
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape filter = context->getInputShape(kFilterTensor);
@@ -572,6 +577,7 @@ bool execute(IOperationExecutionContext* context) {
             NN_RET_CHECK_FAIL() << "Unsupported tensor type for operation " << kOperationName;
     }
 }
+#endif
 
 }  // namespace transpose_conv_2d
 
diff --git a/nn/common/operations/UnidirectionalSequenceLSTM.cpp b/nn/common/operations/UnidirectionalSequenceLSTM.cpp
index f3211692..b74c3c2c 100644
--- a/nn/common/operations/UnidirectionalSequenceLSTM.cpp
+++ b/nn/common/operations/UnidirectionalSequenceLSTM.cpp
@@ -81,6 +81,7 @@ constexpr uint32_t kNumOutputs = 1;
 
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 inline bool hasTensor(IOperationExecutionContext* context, const uint32_t tensor) {
@@ -107,6 +108,7 @@ inline LSTMParams getLSTMParams(IOperationExecutionContext* context) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -156,6 +158,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     // Check that none of the required inputs are omitted
     const std::vector<int> requiredInputs = {
@@ -462,6 +465,7 @@ bool execute(IOperationExecutionContext* context) {
     }
     return true;
 }
+#endif
 
 }  // namespace unidirectional_sequence_lstm
 
diff --git a/nn/common/operations/UnidirectionalSequenceRNN.cpp b/nn/common/operations/UnidirectionalSequenceRNN.cpp
index cf5c620d..c11dba32 100644
--- a/nn/common/operations/UnidirectionalSequenceRNN.cpp
+++ b/nn/common/operations/UnidirectionalSequenceRNN.cpp
@@ -35,6 +35,7 @@ constexpr uint32_t kTimeMajorParam = 6;
 constexpr uint32_t kNumOutputs = 1;
 constexpr uint32_t kOutputTensor = 0;
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 namespace {
 
 template <typename T>
@@ -112,6 +113,7 @@ bool executeTyped(IOperationExecutionContext* context) {
 }
 
 }  // namespace
+#endif
 
 bool validate(const IOperationValidationContext* context) {
     NN_RET_CHECK_EQ(context->getNumInputs(), kNumInputs);
@@ -128,6 +130,7 @@ bool validate(const IOperationValidationContext* context) {
     return validateHalVersion(context, HalVersion::V1_2);
 }
 
+#ifdef NN_INCLUDE_CPU_IMPLEMENTATION
 bool prepare(IOperationExecutionContext* context) {
     Shape input = context->getInputShape(kInputTensor);
     Shape weights = context->getInputShape(kWeightsTensor);
@@ -173,6 +176,7 @@ bool execute(IOperationExecutionContext* context) {
     }
     return true;
 }
+#endif
 
 }  // namespace unidirectional_sequence_rnn
 
-- 
2.17.1

